src/redis.cpp

```diff
@@ -0,0 +1,140 @@
+/**
+* The Forgotten Server - a free and open-source MMORPG server emulator
+* Copyright (C) 2016  Mark Samman <mark.samman@gmail.com>
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License along
+* with this program; if not, write to the Free Software Foundation, Inc.,
+* 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "otpch.h"
+
+#include "redis.h"
+
+Redis::Redis(std::string host, int port, boost::asio::io_service &ioService) :
+	ioService(ioService),
+	publisher(ioService), subscriber(ioService), sync(ioService),
+	address(boost::asio::ip::address::from_string(host)), port(port),
+	scriptInterface("RedisScripts Interface")
+{
+	scriptInterface.initState();
+
+
+	connectSync();
+	connectPublisher();
+	connectSubscriber();
+}
+
+void Redis::stop()
+{
+	stopped = true;
+}
+
+bool Redis::isSuccessfullyInitalized()
+{
+	return syncConnected && publisherConnected && subscriberConnected;
+}
+
+void Redis::connectSync()
+{
+	std::string errmsg;
+
+	if (sync.connect(address, port, errmsg))
+	{
+		syncConnected = true;
+	}
+}
+
+void Redis::connectPublisher()
+{
+	publisher.connect(address, port, [=](bool status, std::string err) {
+		if (status) {
+			publisherConnected = true;
+		}
+	});
+
+	ioService.run_one();
```

Comment:
These two run_one calls are required. Otherwise subscriber and publisher connections won't get estabilished before loading lua scripts that use publish/subscribe method from Redis class and that will cause a crash.